server/src/main/java/org/opensearch/cluster/service/MasterService.java

```diff
@@ -181,15 +183,44 @@ protected void onTimeout(List<? extends BatchedTask> tasks, TimeValue timeout) {
                         )
                     )
                 );
+            String masterThrottlingKey = ((ClusterStateTaskExecutor<Object>) tasks.get(0).batchingKey).getMasterThrottlingKey();
+            masterTaskThrottler.release(masterThrottlingKey, tasks.size());
         }
 
         @Override
         protected void run(Object batchingKey, List<? extends BatchedTask> tasks, String tasksSummary) {
+            // All the batched tasks will have same executor so once they are executed in batch
+            // we will release all batch's permit from master task throttling.
+            String masterThrottlingKey = ((ClusterStateTaskExecutor<Object>) tasks.get(0).batchingKey).getMasterThrottlingKey();
+            masterTaskThrottler.release(masterThrottlingKey, tasks.size());
             ClusterStateTaskExecutor<Object> taskExecutor = (ClusterStateTaskExecutor<Object>) batchingKey;
             List<UpdateTask> updateTasks = (List<UpdateTask>) tasks;
             runTasks(new TaskInputs(taskExecutor, updateTasks, tasksSummary));
         }
 
+        @Override
+        public void submitTasks(List<? extends BatchedTask> tasks, @Nullable TimeValue timeout)
+            throws OpenSearchRejectedExecutionException {
+            assert tasks.size() > 0;
+            String masterThrottlingKey = ((ClusterStateTaskExecutor<Object>) tasks.get(0).batchingKey).getMasterThrottlingKey();
+            if (masterTaskThrottler.acquirePermit(masterThrottlingKey, tasks.size())) {
+                try {
+                    super.submitTasks(tasks, timeout);
+                } catch (Exception e) {
+                    masterTaskThrottler.release(masterThrottlingKey, tasks.size());
+                    throw e;
+                }
```

Comment:
Acquire and release seems too complicated and seems a maintenance overhead

Reply:
Thanks for the suggestion @shwetathareja , 

ConcurrentHashMap's ComputeIfPresent acquires lock on segment of the hashes, so it can create the congestion for multiple task type as for each submit call we will call this to increment the count. Submission of one task type can affect the other task type. This can happen even when throttling is disabled for those task type.

While for Semaphore, it will be on individual task type level only and only if throttling is enabled for that task type.

Do you see any concern with the Semaphores? As with either approach we will need to decrease count/permit from these three places. 

Let me know your thoughts?